src/openvic-simulation/economy/production/ResourceGatheringOperation.cpp

```diff
@@ -14,8 +24,228 @@ ResourceGatheringOperation::ResourceGatheringOperation(
+			}
+		}
+
+		const fixed_point_t base_workforce_size = production_type.get_base_workforce_size();
+		TerrainType const* const terrain_type = location.get_terrain_type();
+		if(terrain_type != nullptr) {
+			ModifierValue const& terrain_modifiers = terrain_type->get_modifier();
+			if(production_type.is_farm()) {
+				//base_workforce_size *= terrain_modifiers["farm_rgo_size"];
+			}
+			if(production_type.is_mine()) {
+				//base_workforce_size *= terrain_modifiers["mine_rgo_size"];
+			}
+		}
+		size_multiplier = ((total_worker_count_in_province / base_workforce_size).ceil() * fixed_point_t::_1_50()).floor();
+	}
+	return total_worker_count_in_province;
+}
+
+void ResourceGatheringOperation::hire(ProvinceInstance& location, Pop::pop_size_t available_worker_count) {
+	total_employees_count_cache = 0;
+	total_paid_employees_count_cache=0;
+	if(this->production_type == nullptr) {
+		employees.empty();
+		return;
+	}
+
+	const Pop::pop_size_t max_worker_count = production_type->get_base_workforce_size() * size_multiplier; // * size modifiers (terrain, province, tech, country, etc)
+	if(max_worker_count <= 0) { return; }
+
+	fixed_point_t proportion_to_hire;
+	if(max_worker_count >= available_worker_count) {
+		//hire everyone
+		proportion_to_hire = fixed_point_t::_1();
+	}
+	else {
+		//hire all pops proportionally
+		const fixed_point_t max_worker_count_real = max_worker_count, available_worker_count_real = available_worker_count;
+		proportion_to_hire = max_worker_count_real / available_worker_count_real;
+	}
+	
+	ProductionType const& production_type = *(this->production_type);
+	std::vector<Job> const& jobs = production_type.get_jobs();
+	std::vector<Pop*> pops = location.get_mutable_pops();
+	for (Pop* pop_nullable : pops){
+		if(pop_nullable == nullptr) {
+			Logger::error("Province ", location.get_identifier(), " has nullptr in get_mutable_pops.");
+			return;
+		}
+		Pop& pop = *pop_nullable;
+		bool is_worker_pop_type = false;
+		PopType const& pop_type = *(pop.get_type());
+		for(Job const& job : jobs) {
+			if(job.get_pop_type() == &pop_type) {
+				is_worker_pop_type = true;
+				break;
+			}
+		}
+
+		if(is_worker_pop_type) {
+			const Pop::pop_size_t pop_size_to_hire = (proportion_to_hire * pop.get_size()).floor().to_int32_t();
+			employees.insert(std::pair<Pop*, Pop::pop_size_t> { &pop, pop_size_to_hire });
+			total_employees_count_cache += pop_size_to_hire;
+			if(!pop_type.get_is_slave()) {
+				total_paid_employees_count_cache += pop_size_to_hire;
+			}
+		}
+	}
+}
+
+fixed_point_t ResourceGatheringOperation::produce(ProvinceInstance& location) {
+	if(this->production_type == nullptr) {
+		return fixed_point_t::_0();
+	}
+	
+	ProductionType const& production_type = *(this->production_type);
+	const fixed_point_t base_workforce_size = production_type.get_base_workforce_size();
+	//TODO include modifiers, job effects, etc
+	return production_type.get_base_output_quantity() * total_employees_count_cache / base_workforce_size;
```

Comment:
Check `base_workforce_size > 0`? Unless we already guaruntee that when loading production types.

Reply:
base_workforce_size is only valid when > 0
I could add checks anyway, but as you pointed out, wasted cpu.
I only write such code when weird crashes make me paranoid.